Brushing in three.js window

richford opened the pull request:
This pull request adds brushing on the three.js viz window. Brush on the FA plots to update the display of the 3d fiber bundles. This pull request also updates three.min.js, adds comments to the 3d-brain.js file. It adds a console message to server.js to tell the user on which port the app will be served. It resamples the fiber tract spatial data so that each tract is the same length (i.e. number of indices) as the FA plot data.

Commit:
Add brushing to FA plots

Commit:
Add console message telling user on which port the app will served

Commit:
Resample fiber tracts to be of same length as FA plot data

Commit:
Update three.min.js to latest version

Commit:
Add .nice() to x domain and reset brush extent on brush deselect

Commit:
Add brushing to three.js window

Commit:
Add read-fibers-convex-hull.ipynb which now resamples the input fiber tracts to be of length 100, same as the FA plot data

Commit:
Merge latest from upstream master

Commit:
Update read-fibers-convex-hull.ipynb which now resamples the input fiber tracts to be of length 100, same as the FA plot data

richford commented:
@arokem, @jyeatman The updated jupyter notebook resamples the tract streamlines using np.interp(). But obviously the three.js part is agnostic about how the data is produced so whatever method you decide to use on the dipy end is fine with me. However the three.js portion does assert that each fiber tract is of the same length as the FA plots. This is something we discussed at the last meeting but if we decide to go with a different data spec, just let me know and I can try to rework it.

jyeatman commented:
I think that makes sense

On Friday, September 16, 2016, Adam Richie-Halford notifications@github.com
wrote:

> @arokem https://github.com/arokem, @jyeatman
> https://github.com/jyeatman The updated jupyter notebook resamples the
> tract streamlines using np.interp(). But obviously the three.js part is
> agnostic about how the data is produced so whatever method you decide to
> use on the dipy end is fine with me. However the three.js portion does
> assert that each fiber tract is of the same length as the FA plots. This is
> something we discussed at the last meeting but if we decide to go with a
> different data spec, just let me know and I can try to rework it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-247746727,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYyiwS5dgaROn9KPGEtAqnGRRRSwOqks5qq2AhgaJpZM4J_foX
> .

## 

Sent from a mobile device. Please excuse typos

arokem commented:
Nice work!

The data specification seems fine to me. I think that it's fair to assume that you will get from our Python processing XYZ coordinates for each streamline that are resampled to 100 points each. 

I fetched this branch from you and played around with this a bit. The interaction with the line-plots works very well -- I think it's quite intuitive. What I am currently seeing is that when I brush on a certain part of an FA map, the gray streamlines for that bundle disappear altogether, and only the part that is brushed is rendered (in the color associated with that bundle). Is that the intended behavior? Another option would be to have the gray streamlines still there, and the colored ones appear on top of that. Thoughts?

richford commented:
@arokem That was the intended behavior but we can definitely do it the way you suggested as well. I could create another branch with that behavior if you want to test these side-by-side. In fact, I was considering an option that goes kind of in the other direction. I was worried that the fiber bundle display was too busy after brushing so I was thinking of just getting rid of all grey (i.e. non-selected) bundles during brushing.

But maybe we should just allow the user to vary the opacity of the non-selected bundles just as we now do with the brain surface view. So this would give us a compromise between the above two approached. The grey streamlines would still be there during brushing, but the user could get rid of them entirely if they felt like they were distracting. The disadvantage would be that it's yet another control for the user to have to worry about. @arokem, @jyeatman, @anotherjoshsmith What do you think?

richford commented:
@arokem, @jyeatman: Also the disadvantage of the data spec with 100 points each is that it increases the file size. Before, we only took every 5th streamline and every 5th point in that streamline. Now, we still take every 5th streamline but we resample to 100 points. So the `data_partial.json` file grew from 16MB to about 30MB. That's per subject if we want to add the ability to do multiple subjects in the 3D-viewer. Not prohibitive but definitely something to consider.

jyeatman commented:
I agree with you suggestion to allow the user to control transparency on
the other fiber groups. Many times it would b nice to make them disappear.
And, in terms of the data file growing, I think that is okay. My
understanding was that switching subjects was going to be done by reading
in a new file right? So we waste disk space but it should cause performance
issues

On Sunday, September 18, 2016, Adam Richie-Halford notifications@github.com
wrote:

> @arokem https://github.com/arokem, @jyeatman
> https://github.com/jyeatman: Also the disadvantage of the data spec
> with 100 points each is that it increases the file size. Before, we only
> took every 5th streamline and every 5th point in that streamline. Now, we
> still take every 5th streamline but we resample to 100 points. So the
> data_partial.json file grew from 16MB to about 30MB. That's per subject
> if we want to add the ability to do multiple subjects in the 3D-viewer. Not
> prohibitive but definitely something to consider.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-247848226,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYyqcDF3oNfwHQ84jIdsWiy62FTkcUks5qrT7dgaJpZM4J_foX
> .

## 

Sent from a mobile device. Please excuse typos

richford commented:
K. I'll push to the brushing branch with these changes:
- Allow user to control opacity of grey streamlines
- Always show grey streamlines "underneath" the colored streamlines

Commit:
Keep grayed-out portion of fiber bundle even while brushing. To implement, we add two bundles to the scene for every one physical bundle. The lighter gray bundles are always on and their opacity is controlled with a slider. The colored bundles are turned on and off by mouseover and selection on the left panel.

richford commented:
I just pushed another commit that add a slider bar for the fiber bundle opacity and keeps the grayed out portion of the fiber bundle as @arokem suggested.

arokem commented:
This is starting to look like a really great added set of functionality. One really useful mode here is to set the fiber opacity to 0 and then highlight just a few bundles, so that only they are visible. Really handy. The main problem I have so far is that the streamlines don't actually become transparent when opacity is set to 0. Instead, they seem to go black, but remain opaque, so that if they are occluding a bundle, you can see them.

Commit:
Change renderOrder and depthWrite properties of all materials so that
transparent line do not occlude visible ones.

See
https://stackoverflow.com/questions/15994944/transparent-objects-in-threejs
for more details. Using depthWrite = true on the gray lines is much
prettier and gives a nice sense of depth (go figure). But we when the
opacity is zero, we need depthWrite = false to solve the issues brought
up in the above referenced stackoverflow question. So we use an if
statement inside the opacity control box to change depthWrite.

richford commented:
@arokem: Good catch. Turns out I stumbled into a well known issue with how WebGL renders transparent objects. I just pushed a commit that works around this to fix the issue you brought up. It's not super elegant but I think it works for our purposes.

All: play around with it and let me know what you think.

jyeatman commented:
Just played around with your branch. It's looking great! What do you think
about adding a check box for "hover over"? I really like the hover over
feature but it would also be nice to turn it off some times

Jason D. Yeatman, PhD
Assistant Professor, Institute for Learning & Brain Sciences (I-LABS)
Department of Speech & Hearing Sciences
University of Washington
http://brainandeducation.com

On Mon, Sep 19, 2016 at 6:53 AM, Adam Richie-Halford <
notifications@github.com> wrote:

> @arokem https://github.com/arokem: Good catch. Turns out I stumbled
> into a well known issue with how WebGL renders transparent objects. I just
> pushed a commit that works around this to fix the issue you brought up.
> It's not super elegant but I think it works for our purposes.
> 
> All: play around with it and let me know what you think.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-247999076,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYyhhY7TPPaPM8iaqtz1v-F8QCop-mks5qrpPKgaJpZM4J_foX
> .

arokem commented:
Looks great! I actually think that we should set default opacity quite low: 0.02 or so looks about right to me. One more small comment: I noticed that there is an interaction between brushing and selecting curves in the line plots. That means that while I am brushing lines get selected and unselected (because the mouse is pressed and we're right on top of them). Is there some way to decouple these, such that if you are brushing, selection is turned off?

Commit:
Add separate stylesheet for the dat.gui controls

Commit:
Add mouseover highlight checkbox to dat.gui controls to toggle mouseover highlighting behavior

Commit:
Update dat.gui.min.js to latest version

richford commented:
@jyeatman I just pushed some commits for toggling the mouseover highlighting on/off. What do you think.

@arokem I'll work on turning off subject selection while brushing.

Commit:
Remove step limitation for brain and fiber opacity sliders and change default fiber opacity to 0.05

Commit:
Prevent brushing and selecting from conflicting on the FA plots

richford commented:
@arokem The latest commits fix the conflict between brushing and selecting. There is still one minor conflict. If you position the cursor directly over a subject's line in the FA plots so that it's mouseover function is activated, and then try to initiate brushing, you'll just wind up selecting a bunch of lines in the FA plot. According to what I've read online, fixing this particular issue is very hard. I can keep looking at it but what we have now resolves most of the issue that you brought up so we might just wind up keeping an open issue for the rest of it.

jyeatman commented:
This looks really good. I played around with the brushing and opacity and
it is working well for me. Once we have the new plotting features up and
running I'm going to have the lab try this out for the current project we
are working on

Jason D. Yeatman, PhD
Assistant Professor, Institute for Learning & Brain Sciences (I-LABS)
Department of Speech & Hearing Sciences
University of Washington
http://brainandeducation.com

On Mon, Sep 19, 2016 at 6:29 PM, Adam Richie-Halford <
notifications@github.com> wrote:

> @arokem https://github.com/arokem The latest commits fix the conflict
> between brushing and selecting. There is still one minor conflict. If you
> position the cursor directly over a subject's line in the FA plots so that
> it's mouseover function is activated, and then try to initiate brushing,
> you'll just wind up selecting a bunch of lines in the FA plot. According to
> what I've read online, fixing this particular issue is very hard. I can
> keep looking at it but what we have now resolves most of the issue that you
> brought up so we might just wind up keeping an open issue for the rest of
> it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-248176806,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYykzeT__lDUPV3dfwtjuFKZXvLJvPks5qrzbjgaJpZM4J_foX
> .

arokem commented on client/js/3d-brain.js:
```diff
@@ -47,9 +83,18 @@ document.body.onmouseup = function() {
+// TODO: Read this from an input file so that the user can change fidelity
+var faPlotLength = 100;
+
+var showStats = true;
```
This variable controls the FPS stats shown on the top left? We probably want to set this back to false before merging, right?

arokem reviewed (commented):
Brushing/selecting works well now. I wouldn't worry about the remaining conflict. 

I don't understand how to toggle the mousover highlighting? Is there supposed to be a checkbox for this? I can't see one (but I might be looking for the wrong thing or in the wrong place).

anotherjoshsmith commented:
The brushing is awesome, good stuff Adam! Working on plots now, sorry for the lack of git integration. I'll get my fork updated before our meeting and smooth out any issues before the weekend.

jyeatman commented:
It's right below the transparency slider

Jason D. Yeatman, PhD
Assistant Professor, Institute for Learning & Brain Sciences (I-LABS)
Department of Speech & Hearing Sciences
University of Washington
http://brainandeducation.com

On Mon, Sep 19, 2016 at 9:22 PM, Ariel Rokem notifications@github.com
wrote:

> _@arokem_ commented on this pull request.
> 
> Brushing/selecting works well now. I wouldn't worry about the remaining
> conflict.
> 
> I don't understand how to toggle the mousover highlighting? Is there
> supposed to be a checkbox for this? I can't see one (but I might be looking
> 
> ## for the wrong thing or in the wrong place).
> 
> In client/js/3d-brain.js
> https://github.com/yeatmanlab/AFQ-viz/pull/13#pullrequestreview-678253:
> 
> > @@ -47,9 +83,18 @@ document.body.onmouseup = function() {
> >    --mouseDown;
> >  }
> 
> -var showStats = false;
> +// Hard coded FA plot data length
> +// TODO: Read this from an input file so that the user can change fidelity
> +var faPlotLength = 100;
> +
> +var showStats = true;
> 
> This variable controls the FPS stats shown on the top left? We probably
> want to set this back to false before merging, right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#pullrequestreview-678253,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYyoKBzPZrfAKG9kCfr1hAdIxPPTiEks5qr1-jgaJpZM4J_foX
> .

Commit:
Turn off stats flag.

richford commented:
@arokem Yep, I just set the stats flag to `false`. Good catch. And the mouseover highlighting checkbox is the third row in the control box underneath the three.js window.

We can choose to have the window expanded or collapsed by default. Since there's some new functionality in there I kept it expanded but if we feel it's distracting we can collapse it by default. What do you think @arokem and @jyeatman?

arokem commented:
Ah yes - there it is! I see the checkbox in Safari, but not in Chrome... Is
that just on my machine?

I think that ultimately we should leave the controls panel closed per
default.

On Tue, Sep 20, 2016 at 5:59 AM, Adam Richie-Halford <
notifications@github.com> wrote:

> @arokem https://github.com/arokem Yep, I just set the stats flag to
> false. Good catch. And the mouseover highlighting checkbox is the third
> row in the control box underneath the three.js window.
> 
> We can choose to have the window expanded or collapsed by default. Since
> there's some new functionality in there I kept it expanded but if we feel
> it's distracting we can collapse it by default. What do you think @arokem
> https://github.com/arokem and @jyeatman https://github.com/jyeatman?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-248293880,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAHPNskxv0g4qFDUHrGIyqY4sKbgryEcks5qr9ilgaJpZM4J_foX
> .

Commit:
close control box by default

richford commented:
I closed the control box by default. The app works as expected for me on Safari, Chrome, and Firefox on my mac. @arokem, can you open the javascript console in Chrome and see if you're getting any errors or warnings?

arokem commented:
Must've been some caching issue. I now see this on Chrome as well. Sorry for the noise. I think this is ready to merge, unless anyone has any more thoughts here?

jyeatman commented:
Yeah, agreed. I have been testing this on chrome and it looks great. Ready
to merge

Jason D. Yeatman, PhD
Assistant Professor, Institute for Learning & Brain Sciences (I-LABS)
Department of Speech & Hearing Sciences
University of Washington
http://brainandeducation.com

On Tue, Sep 20, 2016 at 9:36 AM, Ariel Rokem notifications@github.com
wrote:

> Must've been some caching issue. I now see this on Chrome as well. Sorry
> for the noise. I think this is ready to merge, unless anyone has any more
> thoughts here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/yeatmanlab/AFQ-viz/pull/13#issuecomment-248357734,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABsYypQsRoZ3LWsRUAFuB9ZiKWQJybbtks5qsAuegaJpZM4J_foX
> .
